### PR TITLE
fix(DIST-2017): Race condition with hubspot integration script

### PR DESCRIPTION
**Pull request by mathio**
Closes https://github.com/Typeform/embed/pull/552

This fixes the race condition for when you embed via HTML API (eg `<div data-tf-widget="XX" data-tf-hubspot></div>`). For JavaScript API it is up to the 3rd party engineer to ensure they resolve any race conditions when create their embed.

It does not introduce any breaking changes.

### CODE REVIEW

File: packages/embed/src/initializers/initialize.ts [hunk 1]
```diff
@@ -14,14 +14,18 @@ export const initialize = (
...
 
-  Array.from(embedTypeElements).forEach((element, index) => {
+  for (let index = 0; index < embedTypeElements.length; index += 1) {
+    const element = embedTypeElements.item(index)
     if (forceReload || element.dataset.tfLoaded !== 'true') {
       const formId = element.getAttribute(embedElementAttribute)
       if (!formId) {
         throw new Error(`Invalid ${embedElementAttribute}=${formId} for embed #${index}`)
       }
+      if (element.getAttribute('data-tf-hubspot') !== null) {
```
Comment (maxprilutskiy) [inline packages/embed/src/initializers/initialize.ts:24]: how about `undefined`? Maybe worth changing to `!= null` to cover `undefined` as well?

**Comment by typeform-ops-gha** [comment]
*[BOT]* Preview available with hash `57f5653b3bc2248e268045297efd394e3ed4cba7` [here](https://embed.typeform.com/57f5653b3bc2248e268045297efd394e3ed4cba7/embed.js).

File: packages/embed/src/utils/hubspot.ts [hunk 2]
```diff
@@ -6,23 +8,19 @@ type HubspotFieldsType = {
...
- */
-
 export const getHubspotHiddenFields = (): HubspotFieldsType => ({
   hubspot_page_name: document.title,
   hubspot_page_url: window.location.href,
   hubspot_utk: getHubspotCookieValue(),
 })
+
+export const waitForHubspotCookie = async (): Promise<void> => {
+  for (let maxRetries = 10, retries = 0; !getHubspotCookieValue() && retries < maxRetries; retries++) {
```
Comment (maxprilutskiy) [inline packages/embed/src/utils/hubspot.ts:23]: It looks like `getHubspotCookieValue()` will get called on every iteration, during the for condition evaluation, is that the idea? It's ok if it is, just thought I'd mention

**Review by maxprilutskiy — APPROVED** [review]

File: packages/embed/src/initializers/initialize.ts (hunk 1 above)
Comment (hburbano) [reply packages/embed/src/initializers/initialize.ts:24]:
```suggestion
      if (element.getAttribute('data-tf-hubspot')) {
```
An alternative here could be checking straightforward for a [Truthy](https://developer.mozilla.org/en-US/docs/Glossary/Truthy) value

Comment (mathio) [reply packages/embed/src/initializers/initialize.ts:24]: When the attribute is not present the [return value is `null`](https://developer.mozilla.org/en-US/docs/Web/API/Element/getAttribute#non-existing_attributes). However if it is present, the value is an empty string and therefore [Falsy](https://developer.mozilla.org/en-US/docs/Glossary/Falsy) :D

Such is a life with JavaScript and browser APIs 😛

File: packages/embed/src/utils/hubspot.ts (hunk 2 above)
Comment (mathio) [reply packages/embed/src/utils/hubspot.ts:23]: Yes, thats the idea. It is a (hopefully quick) regex check. We get to break out from the "sleep cycle" as soon as the cookie is present.

Comment (mrisoli) [reply packages/embed/src/utils/hubspot.ts:23]: Could this be done with [cookies.onChanged?](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/onChanged)

Comment (mathio) [reply packages/embed/src/utils/hubspot.ts:23]: @mrisoli I see it is currently [not supported in Safari](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/cookies/onChanged#browser_compatibility) 😢

**Comment by typeform-ops-gha** [comment]
*[BOT]* Preview available with hash `40dfab5c5524cc6efcbf82f702208d4c8bb9c65d` [here](https://embed.typeform.com/40dfab5c5524cc6efcbf82f702208d4c8bb9c65d/embed.js).

**Review by mrisoli — APPROVED** [review]

**Pull request merged**